### PR TITLE
executor,planner: Improve SET PASSWORD

### LINKED ISSUE #7522: set password='xx' fails privilege check
Please answer these questions before submitting your issue. Thanks!

1. What did you do?

I tried to change the password for a non-privileged user:
```
mysql> create user nobody;
^c
shell> mysql -unobody                                                                                                                                                                                 mysql> set password='acdc';
ERROR 1105 (HY000): privilege check fail
mysql> SET password=password('acdc');                                                                                                                                                                            ERROR 1105 (HY000): privilege check fail
```

2. What did you expect to see?

It should be possible to set your own password.

3. What did you see instead?

Error.

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

```
MySQL [(none)]> SELECT tidb_version()\G                                                                                                                                                                                   *************************** 1. row ***************************
tidb_version(): Release Version: v2.1.0-rc.1-13-g1255973
Git Commit Hash: 1255973764f72575d71920b95fc0861d1677c6c5
Git Branch: master
UTC Build Time: 2018-08-28 07:35:57
GoVersion: go version go1.10.2 linux/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by morgo**
### What problem does this PR solve? 

Fixes #7522

### What is changed and how it works?

* Set password was too strict in privilege checking, and required SUPER for all use cases.  Now it only requires SUPER for changing other user's passwords.
* Set password did not use the AuthUsername + AuthPassword session properties.  So if a user account was a wildcard, it was not possible to change the password.
* Error messages/codes are now mysql compatible.

This PR will conflict with #8417 as they both add the same error to the executor.  I will see which one is merged first and then fix the code :-)

There remains an issue common to all GRANT/SET password/CREATE user commands in that they do not invalidate the privilege cache (as MySQL versions do).  See https://github.com/pingcap/tidb/issues/8427

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)

Code changes

 - minimal

Side effects

 - Breaking backward compatibility (broken apps)

Related changes

 - Need to be included in the release note

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8426)

### CODE REVIEW

**Comment by morgo** [comment]
PTAL @tiancaiamao

**Comment by tiancaiamao** [comment]
LGTM
Good job! @morgo

File: executor/simple.go [hunk 1]
```diff
@@ -287,14 +288,22 @@ func userExists(ctx sessionctx.Context, name string, host string) (bool, error)
 }
 
 func (e *SimpleExec) executeSetPwd(s *ast.SetPwdStmt) error {
+	var u, h string
 	if s.User == nil {
-		vars := e.ctx.GetSessionVars()
-		s.User = vars.User
-		if s.User == nil {
+		if e.ctx.GetSessionVars().User == nil {
 			return errors.New("Session error is empty")
```
Comment (jackysp) [inline executor/simple.go]: Why the error is "Session error is empty"? Although it is not related to this PR.

**Pull request merged**